# google-covid-19-mobility-data-process-usa

In this notebook I will refine the process from `google-covid-19-mobility-data-process-v1` in order to get the highest resolution data possible for a map of the USA.

In [1]:
import pandas as pd
import requests
import simplejson as json
import numpy as np

---

## Load a reduced CSV containing just the United States entries

In [2]:
usaDf = pd.read_csv("./output-data/usa.csv", low_memory=False)

In [3]:
usaDf.head()

country_region_code country_region sub_region_1 sub_region_2  metro_area  \
0                  US  United States          NaN          NaN         NaN   
1                  US  United States          NaN          NaN         NaN   
2                  US  United States          NaN          NaN         NaN   
3                  US  United States          NaN          NaN         NaN   
4                  US  United States          NaN          NaN         NaN   

  iso_3166_2_code  census_fips_code                     place_id        date  \
0             NaN               NaN  ChIJCzYy5IS16lQRQrfeQ5K5Oxw  2020-02-15   
1             NaN               NaN  ChIJCzYy5IS16lQRQrfeQ5K5Oxw  2020-02-16   
2             NaN               NaN  ChIJCzYy5IS16lQRQrfeQ5K5Oxw  2020-02-17   
3             NaN               NaN  ChIJCzYy5IS16lQRQrfeQ5K5Oxw  2020-02-18   
4             NaN               NaN  ChIJCzYy5IS16lQRQrfeQ5K5Oxw  2020-02-19   

   retail_and_recreation_percent_change_from_baseline  \
0                                                6.0    
1                                                7.0    
2                                                6.0    
3                                                0.0    
4                                                2.0    

   grocery_and_pharmacy_percent_change_from_baseline  \
0                                                2.0   
1                                                1.0   
2                                                0.0   
3                                               -1.0   
4                                                0.0   

   parks_percent_change_from_baseline  \
0                                15.0   
1                                16.0   
2                                28.0   
3                                 6.0   
4                                 8.0   

   transit_stations_percent_change_from_baseline  \
0                                            3.0   
1                                            2.0   
2                                           -9.0   
3                                            1.0   
4                                            1.0   

   workplaces_percent_change_from_baseline  \
0                                      2.0   
1                                      0.0   
2                                    -24.0   
3                                      0.0   
4                                      1.0   

   residential_percent_change_from_baseline  
0                                      -1.0  
1                                      -1.0  
2                                       5.0  
3                                       1.0  
4                                       0.0

---

**`sub_reigon_1` is states**

In [9]:
usaDf["sub_region_1"].unique()

array([nan, 'Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

---

**`sub_region_2` is counties - of which there are 1,719**

In [10]:
usaDf["sub_region_2"].unique()

array([nan, 'Autauga County', 'Baldwin County', ..., 'Uinta County',
       'Washakie County', 'Weston County'], dtype=object)

In [11]:
len(usaDf["sub_region_2"].unique())

1719

---

**`metro_area` is not used**

In [13]:
usaDf["metro_area"].unique()

array([nan])

---

**`iso_3166_2_code` is used and relates to States.**

In [14]:
usaDf["iso_3166_2_code"].unique()

array([nan, 'US-AL', 'US-AK', 'US-AZ', 'US-AR', 'US-CA', 'US-CO', 'US-CT',
       'US-DE', 'US-DC', 'US-FL', 'US-GA', 'US-HI', 'US-ID', 'US-IL',
       'US-IN', 'US-IA', 'US-KS', 'US-KY', 'US-LA', 'US-ME', 'US-MD',
       'US-MA', 'US-MI', 'US-MN', 'US-MS', 'US-MO', 'US-MT', 'US-NE',
       'US-NV', 'US-NH', 'US-NJ', 'US-NM', 'US-NY', 'US-NC', 'US-ND',
       'US-OH', 'US-OK', 'US-OR', 'US-PA', 'US-RI', 'US-SC', 'US-SD',
       'US-TN', 'US-TX', 'US-UT', 'US-VT', 'US-VA', 'US-WA', 'US-WV',
       'US-WI', 'US-WY'], dtype=object)

---

**`census_fips_code` is present and could be used as an alternative to `place_id`**

In [15]:
usaDf["census_fips_code"].unique()

array([   nan,  1001.,  1003., ..., 56041., 56043., 56045.])

---

## Create a new data frame containing just the rows with `sub_region_2`

In [6]:
len(usaDf["place_id"].unique())

2889

In [7]:
usaSubRegion2Df = usaDf[usaDf["sub_region_2"].notnull()]

In [8]:
len(usaSubRegion2Df["place_id"].unique())

2837

We now have 2,837 rows instead of 2,889. Not a great reduction...

---

## Get the lat long coordinates for each unique `place_id`

In [29]:
uniquePlaceIdsDf = usaSubRegion2Df[["place_id"]].drop_duplicates()

Access the google maps api to get coordinates for each `place_id`

In [30]:
with open('./secrets/googleapikey.txt', 'r') as f:
    key = f.read()

In [31]:
def get_lat_long(place_id):
    try:
        API_KEY = key.rstrip("\n")
        url = "https://maps.googleapis.com/maps/api/place/details/json?place_id=" + str(place_id) + "&key=" + str(API_KEY) + "&fields=geometry"
        
        response = (requests.get(url).text)
        response_json = json.loads(response)
        
        if "result" in response_json:
            result = response_json["result"]
            if "geometry" in result:
                geometry = result["geometry"]
                if "location" in geometry:
                    location = geometry["location"]
                    return location["lat"], location["lng"]
                else:
                    return None, None
            else:
                return None, None
        else:
            return None, None
    
    except Exception as e:
        raise e

In [32]:
uniquePlaceIdsDf.loc[:, "lat"], uniquePlaceIdsDf.loc[:, "lng"] = zip(*uniquePlaceIdsDf['place_id'].map(get_lat_long))

In [33]:
uniquePlaceIdsDf.head()

place_id        lat        lng
924   ChIJg9z7ewWPjogRA_8QrB0va7o  32.579182 -86.499655
1380  ChIJMRERvxBnmogR0K_9O08dxDw  30.601074 -87.776333
1842  ChIJtUJfgR1tjYgRAjFO7yq2nuI  31.817290 -85.354965
2279  ChIJEdvIA8szj4gRU_XpIL_1cSw  32.956280 -87.142289
2716  ChIJdcUJhYeXiYgRERpotPlZFfs  34.014515 -86.499655

I'll save these coordinates as a CSV file for later use.

In [34]:
uniquePlaceIdsDf.to_csv("./output-data/usa-sub-region-2-lat-lng.csv", index=False)

In [35]:
len(uniquePlaceIdsDf)

2837

---

## Merge the coordinates with the original `sub_region_2` data frame

In [4]:
uniquePlaceIdsDf = pd.read_csv("./output-data/usa-sub-region-2-lat-lng.csv", low_memory=False)

In [9]:
usaSubRegion2MergeDf = pd.merge(usaSubRegion2Df, uniquePlaceIdsDf, on='place_id', how='outer')

In [10]:
usaSubRegion2MergeDf.head()

country_region_code country_region sub_region_1    sub_region_2  metro_area  \
0                  US  United States      Alabama  Autauga County         NaN   
1                  US  United States      Alabama  Autauga County         NaN   
2                  US  United States      Alabama  Autauga County         NaN   
3                  US  United States      Alabama  Autauga County         NaN   
4                  US  United States      Alabama  Autauga County         NaN   

  iso_3166_2_code  census_fips_code                     place_id        date  \
0             NaN            1001.0  ChIJg9z7ewWPjogRA_8QrB0va7o  2020-02-15   
1             NaN            1001.0  ChIJg9z7ewWPjogRA_8QrB0va7o  2020-02-16   
2             NaN            1001.0  ChIJg9z7ewWPjogRA_8QrB0va7o  2020-02-17   
3             NaN            1001.0  ChIJg9z7ewWPjogRA_8QrB0va7o  2020-02-18   
4             NaN            1001.0  ChIJg9z7ewWPjogRA_8QrB0va7o  2020-02-19   

   retail_and_recreation_percent_change_from_baseline  \
0                                                5.0    
1                                                0.0    
2                                                8.0    
3                                               -2.0    
4                                               -2.0    

   grocery_and_pharmacy_percent_change_from_baseline  \
0                                                7.0   
1                                                1.0   
2                                                0.0   
3                                                0.0   
4                                                0.0   

   parks_percent_change_from_baseline  \
0                                 NaN   
1                               -23.0   
2                                 NaN   
3                                 NaN   
4                                 NaN   

   transit_stations_percent_change_from_baseline  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

   workplaces_percent_change_from_baseline  \
0                                     -4.0   
1                                     -4.0   
2                                    -27.0   
3                                      2.0   
4                                      2.0   

   residential_percent_change_from_baseline        lat        lng  
0                                       NaN  32.579182 -86.499655  
1                                       NaN  32.579182 -86.499655  
2                                       5.0  32.579182 -86.499655  
3                                       0.0  32.579182 -86.499655  
4                                       0.0  32.579182 -86.499655

---

## Calculate 7 day rolling averages for each location

In [11]:
def add_rolling_average(df):
    df.loc[:, "retail-average"] = df["retail_and_recreation_percent_change_from_baseline"].rolling(window=7, center=True).mean()
    df.loc[:, "grocery-average"] = df["grocery_and_pharmacy_percent_change_from_baseline"].rolling(window=7, center=True).mean()
    df.loc[:, "parks-average"] = df["parks_percent_change_from_baseline"].rolling(window=7, center=True).mean()
    df.loc[:, "transit-average"] = df["transit_stations_percent_change_from_baseline"].rolling(window=7, center=True).mean()
    df.loc[:, "workplace-average"] = df["workplaces_percent_change_from_baseline"].rolling(window=7, center=True).mean()
    df.loc[:, "residential-average"] = df["residential_percent_change_from_baseline"].rolling(window=7, center=True).mean()
    
    return df

In [12]:
usaSubRegion2AverageDf = usaSubRegion2MergeDf.groupby("place_id").apply(add_rolling_average)

In [13]:
usaSubRegion2AverageDf[1000000:1000005]

country_region_code country_region sub_region_1   sub_region_2  \
1000000                  US  United States        Texas  Taylor County   
1000001                  US  United States        Texas  Taylor County   
1000002                  US  United States        Texas  Taylor County   
1000003                  US  United States        Texas  Taylor County   
1000004                  US  United States        Texas  Taylor County   

         metro_area iso_3166_2_code  census_fips_code  \
1000000         NaN             NaN           48441.0   
1000001         NaN             NaN           48441.0   
1000002         NaN             NaN           48441.0   
1000003         NaN             NaN           48441.0   
1000004         NaN             NaN           48441.0   

                            place_id        date  \
1000000  ChIJNfnP60m-VoYRN7ziFE_S9ic  2021-02-09   
1000001  ChIJNfnP60m-VoYRN7ziFE_S9ic  2021-02-10   
1000002  ChIJNfnP60m-VoYRN7ziFE_S9ic  2021-02-11   
1000003  ChIJNfnP60m-VoYRN7ziFE_S9ic  2021-02-12   
1000004  ChIJNfnP60m-VoYRN7ziFE_S9ic  2021-02-13   

         retail_and_recreation_percent_change_from_baseline  ...  \
1000000                                              -12.0   ...   
1000001                                              -23.0   ...   
1000002                                              -38.0   ...   
1000003                                              -24.0   ...   
1000004                                              -39.0   ...   

         workplaces_percent_change_from_baseline  \
1000000                                    -19.0   
1000001                                    -22.0   
1000002                                    -46.0   
1000003                                    -27.0   
1000004                                    -24.0   

         residential_percent_change_from_baseline        lat        lng  \
1000000                                       6.0  32.245509 -99.812494   
1000001                                      10.0  32.245509 -99.812494   
1000002                                      21.0  32.245509 -99.812494   
1000003                                      14.0  32.245509 -99.812494   
1000004                                      12.0  32.245509 -99.812494   

         retail-average  grocery-average  parks-average  transit-average  \
1000000      -19.714286        -6.428571            NaN         4.428571   
1000001      -23.857143        -7.857143            NaN         1.571429   
1000002      -32.142857       -14.428571            NaN        -5.142857   
1000003      -39.142857       -17.142857            NaN       -12.857143   
1000004      -47.000000       -17.428571            NaN       -24.142857   

         workplace-average  residential-average  
1000000         -22.000000             9.142857  
1000001         -23.857143            10.285714  
1000002         -29.571429            12.428571  
1000003         -38.571429            15.142857  
1000004         -47.142857            16.857143  

[5 rows x 23 columns]

---

## Round the average figures to 1 decimal place for a smaller final file size

In [14]:
usaSubRegion2RoundedDf = usaSubRegion2AverageDf.round({
    'retail-average': 1,
    'grocery-average': 1,
    'parks-average': 1,
    'transit-average': 1,
    'workplace-average': 1,
    'residential-average': 1
})

---

## Remove any NaN `place_id`s

In [15]:
usaSubRegion2NotNaDf = usaSubRegion2RoundedDf[usaSubRegion2RoundedDf["place_id"].notna(
)]

---

## Convert the data into a python dictionary so it can be exported as json

In [22]:
def create_list_for_json(df):
    outputList = []
    listOfPlaceIds = df["place_id"].drop_duplicates().to_list()
    groupByPlaceId = df.groupby("place_id")
    
    for place_id in listOfPlaceIds:
        thisDf = groupByPlaceId.get_group(place_id)
        parksList = thisDf["parks-average"].to_list()
        
        # Some of the parks columns contain all NaNs, we'll skip these
        if np.isnan(parksList).all():
            continue
        
        myDict = {}
        myDict['lng'] = thisDf.iloc[0]["lng"]
        myDict["lat"] =  thisDf.iloc[0]["lat"]

        # parks_percent_change_from_baseline
        myDict["parks"] = thisDf.set_index("date")["parks-average"].to_dict()

        outputList.append(myDict)
        
    return outputList

In [23]:
usaSubRegion2List = create_list_for_json(usaSubRegion2NotNaDf)

In [25]:
usaSubRegion2List[0]

{'lng': -87.77633329999998,
 'lat': 30.6010744,
 'parks': {'2020-02-15': nan,
  '2020-02-16': nan,
  '2020-02-17': nan,
  '2020-02-18': 26.4,
  '2020-02-19': 26.6,
  '2020-02-20': 41.4,
  '2020-02-21': 38.7,
  '2020-02-22': 40.6,
  '2020-02-23': 40.1,
  '2020-02-24': 47.1,
  '2020-02-25': 48.1,
  '2020-02-26': 55.7,
  '2020-02-27': 50.1,
  '2020-02-28': 52.3,
  '2020-02-29': 52.7,
  '2020-03-01': 48.1,
  '2020-03-02': 46.1,
  '2020-03-03': 45.7,
  '2020-03-04': 39.9,
  '2020-03-05': 42.6,
  '2020-03-06': 45.0,
  '2020-03-07': 48.3,
  '2020-03-08': 56.3,
  '2020-03-09': 62.0,
  '2020-03-10': 68.0,
  '2020-03-11': 71.4,
  '2020-03-12': 73.1,
  '2020-03-13': 73.7,
  '2020-03-14': 73.3,
  '2020-03-15': 72.7,
  '2020-03-16': 67.9,
  '2020-03-17': 54.0,
  '2020-03-18': 37.3,
  '2020-03-19': 23.3,
  '2020-03-20': 12.6,
  '2020-03-21': 1.4,
  '2020-03-22': -8.7,
  '2020-03-23': -16.3,
  '2020-03-24': -18.3,
  '2020-03-25': -19.3,
  '2020-03-26': -21.1,
  '2020-03-27': -22.6,
  '2020-03-28': -2

In [24]:
len(usaSubRegion2List)

666

In [26]:
with open("./public/data/usa-parks.json", "w") as outfile: 
    json.dump(usaSubRegion2List, outfile, ignore_nan=True)